### PR TITLE
Distributed multi-observation mapmaking (v2)

### PR DESCRIPTION
**Pull request by sbiquard**
Like #107 but using new block operators to fuse operations in scan loops.

Manual parallelism (shard_map) is needed because we want each process to use `scan` when looping through its observations (for memory reasons), and the inputs to `scan` must be fully replicated. In other words, each process should only see its own "local" slice of the data, which is what shard_map provides.

The new `ScanBlock{...}Operator` classes provide a convenient abstraction around the pattern where an operator pytree has been stacked along an axis (typically for observations) and needs to be applied to a stacked input vector using `scan`. Those operators do not embed sharding information in their input/output structures, and use sharding purely at runtime via `jax.reshard` and `jax.shard_map` inside `mv()`. This prevents structure equality checks (when composing operators) from failing in various places since not all the codebase is sharding-friendly/compatible.

Similarly, calls to shard_map use `check_vma=False` because we rely on `jax.linear_transpose` which traces the matvec function from the operator's input structure, constructed outside the shard and therefore lacking varying-axis (vma) annotations. This results in a mismatch between the traced function output and the cotangent type, which IS varying at call time.

`MapMakingConfig` and all its components have been made frozen dataclasses so that they are hashable and be passed as static arguments to jitted functions (namely `accumulate_rhs` in this case).

### CODE REVIEW

File: src/furax/mapmaking/_scan_blocks.py [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+
+def _get_mesh() -> AbstractMesh:
+    mesh = jax.sharding.get_abstract_mesh()
+    if mesh.empty:
+        raise RuntimeError('active mesh context required')
+    return mesh
+
+
+class AbstractScanBlockOperator(AbstractLinearOperator, ABC):
+    blocks: AbstractLinearOperator
```
Comment (Wuhyun) [inline src/furax/mapmaking/_scan_blocks.py]: Are there any assumptions made for these 'blocks' operators? It would help me understand the code better!

File: src/furax/mapmaking/_scan_blocks.py [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+
+    @property
+    def out_structure(self) -> PyTree[jax.ShapeDtypeStruct]:
+        n = jax.tree.leaves(self.in_structure)[0].shape[0]
+        return _prepend_axis(self.blocks.out_structure, axis_size=n)
+
+    def mv(self, x: PyTree[Inexact[Array, '...']]) -> PyTree[Inexact[Array, '...']]:
+        axis = _get_mesh().axis_names[0]
+
+        @jax.shard_map(out_specs=P(axis), check_vma=False)
```
Comment (Wuhyun) [inline src/furax/mapmaking/_scan_blocks.py:101]: I don't know much about this `check_vma` argument, but it looks like it's False by default, and the [JAX documentation](https://docs.jax.dev/en/latest/config_options.html#check_vma) says 'DO NOT USE'?

File: src/furax/mapmaking/_scan_blocks.py (hunk 1 above)
Comment (sbiquard) [reply src/furax/mapmaking/_scan_blocks.py]: Good point, let me add some docstrings

File: src/furax/mapmaking/_scan_blocks.py (hunk 2 above)
Comment (sbiquard) [reply src/furax/mapmaking/_scan_blocks.py:101]: I didn't know it was a global configuration option. It is a valid parameter per the [shard_map documentation](https://docs.jax.dev/en/latest/_autosummary/jax.shard_map.html) and defaults to True.

File: src/furax/mapmaking/_scan_blocks.py [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+                op, x_i = args
+                return None, op.mv(x_i)
+
+            _, out = jax.lax.scan(step, None, (blocks, x))
+            return out
+
+        # shard_map validates in_specs against actual array sharding; reshard sets it
+        # explicitly so the check passes even when sharding is lost inside a JIT trace
+        blocks = jax.tree.map(lambda a: jax.reshard(a, P(axis)), self.blocks)
+        x = jax.tree.map(lambda a: jax.reshard(a, P(axis)), x)
```
Comment (Wuhyun) [inline src/furax/mapmaking/_scan_blocks.py]: These `reshard`s are quite confusing - what do they do exactly? Are we resharding it to the same sharding just to override some checks?...

Comment (sbiquard) [reply src/furax/mapmaking/_scan_blocks.py]: It turns out they don't seem required in the end, but rather left over from previous iterations on this PR. I'll remove them!
